Fixes make check, and adding a doc check on Gh actions :books:

vincenzopalazzo opened the pull request:
Fixes #4773
Fixes #4904

I think that this is the cause of make check failure, but this command continues to fail. I'm not able to fully understand the make script, but maybe @rustyrussell can give me some tips?

- [X] Fix `make check` command;
- [X] Adding a not dependent check of docs in the GH to make safe also the merge operation.
~- [ ] Adding missing parameters well documented in the `lightningd.config.5.md`~
~- [ ] Adding missing parameters (that was added in the previous point) into the `jsonschema`.~
- [X] Skip experimental proprieties

Signed-off-by: Vincenzo Palazzo <vincenzopalazzodev@gmail.com>

vincenzopalazzo commented:
In my opinion, the human is very bad at formatting stuff, and I preferer some tools that are able to do it each time with the same provision. With the JSON schema, this time was easy.

I remove the space check in the Makefile and create another tool to formatting the JSON schema. Now should be more easy write (formatting) the JSON schema and make all happy.

I introduced also the check-in to the CI, to make sure that we merge only documented commands.


Only a small question here with the following error https://github.com/vincenzopalazzo/lightning/runs/3572055136?check_suite_focus=true#step:5:1572 with the command `make -j$(nproc) check-manpages`, I think I'm missing something very small, I will look into it with a fresh mind tomorrow.

Thanks

jsarenik commented:
@vincenzopalazzo the patch is _huge_ and seems to be editing white-space in unrelated files.

When I do just following to current `master` (cdb93eefe) …

```diff
diff --git a/doc/schemas/listconfigs.schema.json b/doc/schemas/listconfigs.schema.json
index 877cea210..96f0e40a4 100644
--- a/doc/schemas/listconfigs.schema.json
+++ b/doc/schemas/listconfigs.schema.json
@@ -243,6 +243,10 @@
 	    "type": "string",
 	    "description": "`subdaemon` fields from config or cmdline if any (can be more than one)"
 	},
+	"fetchinvoice-noconnect": {
+	    "type": "string",
+	    "description": "prevents `fetchinvoice` and `sendinvoice` from trying to connect directly"
+	},
 	"tor-service-password": {
 	    "type": "string",
 	    "description": "`tor-service-password` field from config or cmdline, if any"
```

… the first error\* is gone but then I get another error later in the end:

```
+ tools/check-manpage.sh lightningd/lightningd --lightning-dir=/tmp/ doc/lightningd-config.5.md
diff of command names vs manpage names:
--- /dev/fd/63	2021-09-12 14:06:41.419619727 +0200
+++ /dev/fd/62	2021-09-12 14:06:41.419619727 +0200
@@ -14,8 +14,6 @@
 bitcoin-rpcpassword=
 bitcoin-rpcport=
 bitcoin-rpcuser=
-channel-fee-max-base-msat=
-channel-fee-max-proportional-thousandths=
 clear-plugins
 cltv-delta=
 cltv-final=
@@ -35,24 +33,11 @@
 fee-per-satoshi=
 fetchinvoice-noconnect
 force-feerates=
-funder-fund-probability=
-funder-fuzz-percent=
-funder-lease-requests-only=
-funder-max-their-funding=
-funder-min-their-funding=
-funder-per-channel-max=
-funder-per-channel-min=
-funder-policy-mod=
-funder-policy=
-funder-reserve-tank=
 funding-confirms=
 help
 ignore-fee-limits=
 important-plugin=
 large-channels
-lease-fee-base-msat=
-lease-fee-basis=
-lease-funding-weight=
 lightning-dir=
 log-file=
 log-level=
```

… which could be the reason for some other other edits I guess, but not 244 files. What do you think?

\* the original error which gets fixed by the little patch above like looks this:

```
$ make check                                                                   
fetchinvoice-noconnect documented but not in schema!
make: *** [doc/Makefile:152: check-config-docs] Error 1
```

vincenzopalazzo commented:
> which could be the reason for some other edits I guess, but not 244 files. What do you think?

I think that the `lightningd-config.5.md`, doesn't contain all the real propriety in the doc.

The real change inside this PR is
- Create formatting tools in python, that take the schemas and put them in a standard format ([see](https://github.com/ElementsProject/lightning/pull/4782/files#diff-d4e516f7fe801ab2391caaf857f40aea661b2e2c1c933d1b4a12aa96f5366ac6)) you can see also that are some formatting fixes derived from some copy and paste operations.
- Add your change in the JSON schema
- Remove the checking of the space in the make file, because with the tools we have an easy life
- Put the doc check in the Github action
- Run make doc-all that will generate a ton of files from the MD file where the only difference is the hash


This PR is more related to make sure that we are not error derived from space during the checking of the doc with the schema, and make sure that we will not merge anymore `undoc` PR (at list without knowing it)

I'm sorry if my change will generate a ton of changes in the PR

rustyrussell commented:
Can't we just use jq for formatting?  i.e. `jq < orig > new`?

vincenzopalazzo commented:
> Can't we just use jq for formatting? i.e. jq < orig > new?

Before ti fall into python I tried to use jq too, but I failed. I used the following command `echo $(cat offerout.schema.json | jq .) > config.json` but the JSON was unformatted (on in one line). Suggestion on how to fix it?

Thanks

cdecker commented:
Welcome to the wonderful world of checking in derived files...

The main issue is that since the `SHA256STAMP` seems to cover some shared state, a change in any of the schemas will trigger _all_ doc md files and therefore _all_ manpages to be regenerated.

jsarenik commented:
> > Can't we just use jq for formatting? i.e. jq < orig > new?
> 
> Before ti fall into python I tried to use jq too, but I failed. I used the following command `echo $(cat offerout.schema.json | jq .) > config.json` but the JSON was unformatted (on in one line). Suggestion on how to fix it?

Hi @vincenzopalazzo, just do

```
cat offerout.schema.json | jq . > config.json
```

instead of shell-evaluating it inside `$(...)`.

vincenzopalazzo commented:
Thank @jsarenik, I make only one change in the command

Rebased https://github.com/vincenzopalazzo/lightning/pull/4782/commits/0cbf95b0b419fdadd0eb255bce71b7540377ded5

- Remove the python tool
- use `jq` to formatting the `json` with a new make target, (maybe we need to put jq in the install section of the doc?)
- formatting in each make doc-all

Still missing the fix on json-schema but I'm working around on it

jsarenik reviewed (approved):
Looks good to me now @vincenzopalazzo . Now I get the big amount of generated changes.

ACK oops, update: not so fast,: the added CI doc check needs mako which is not installed by default

jsarenik commented:
ACK 0cbf95b0b419fdadd0eb255bce71b7540377ded5

jsarenik commented on .github/workflows/ci.yaml:
```diff
@@ -66,6 +66,31 @@ jobs:
+        run: bash -x .github/scripts/setup.sh
+
+      - name: Check Manpages
+        run: |
```
Add a line `pip3 install mako` here

jsarenik replied:
I meant right after `run: |`, but could be elsewhere maybe. Didn't test it.

cdecker replied:
No, that'd fail in the `NO_PYTHON` configuration which is run on CI. Incidentally this should rather be fixed by checking in the `channel_type_wiregen` files in #4794

jsarenik reviewed (changes requested):
Mako is not installed by default, please make sure it is installed (does not have to be exactly where I commented, feel free to do it elsewhere).

jsarenik commented:
This draft commit may need a rebase if https://github.com/ElementsProject/lightning/pull/4794 gets merged.

vincenzopalazzo commented on doc/lightningd-config.5:
```diff
@@ -328,6 +328,66 @@ values), and the order is "opening", "mutual_close", "unilateral_close",
+
+
+ \fBfunder-max-their-funding\fR==\fIMILLIONTHS\fR
+adding explanation and type
+
+
+ \fBfunder-min-their-funding\fR==\fIMILLIONTHS\fR
+adding explanation here and type info
+
+
+ \fBfunder-per-channel-max\fR==\fIMILLIONTHS\fR
+adding explanation here and type info
+
+
+ \fBfunder-per-channel-min\fR==\fIMILLIONTHS\fR
+adding explanation here and type info
+
+
+ \fBfunder-policy-mod\fR==\fIMILLIONTHS\fR
+adding explanation here and type info
+
+
+ \fBfunder-policy\fR==\fIMILLIONTHS\fR
+adding explanation here and type info
+
+
+ \fBfunder-reserve-tank\fR==\fIMILLIONTHS\fR
+adding explanation here and type info
+
+
+ \fBlease-fee-base-msat\fR==\fIMILLIONTHS\fR
+adding explanation here and type info
+
+
+ \fBlease-fee-basis\fR==\fIMILLIONTHS\fR
+adding explanation here and type info
+
+
+ \fBlease-funding-weight\fR==\fIMILLIONTHS\fR
+adding explanation here and type info
```
Removed from the doc. Thanks

vincenzopalazzo commented on doc/Makefile:
```diff
@@ -89,7 +89,13 @@ MANPAGES := doc/lightning-cli.1 \
+#FIXME(vincenzopalazzo) we don't need to change a file if it is well format,
+#so we can use diff to check difference in the real json and in the .fmt
+fmt-schema:
+	@for f in $$(find doc/schemas -type f -name '*.json'); do cat "$$f" | jq --tab . > "$$f.fmt" ; mv "$$f.fmt" "$$f" ; done
```
Happy Spaces 🛩️

vincenzopalazzo commented:
Rebased on top of master https://github.com/ElementsProject/lightning/pull/4782/commits/be3f21506a056ad89427d4b381ad563de2a78044  includes:

1. Move from tab to space 🛩️ 
2. Removed experimental feature from the `lightningd-conf.md`
3. check-manpace `still failing, we need to pull out the experimental stuff out of check.

The proposal to fix 3 can be:
filter by the experimental-* prefix, support the actual experimental stuff without checking the prefix, but look inside a json file, that contains all the proprieties listed, like below:

```
{
     "name_prop": true
}
```

The json file gives us the flexibility to make the migration of the name proprieties later, and when the migration will happen we can remove this json file

vincenzopalazzo commented:
Here we go, I have the first complete version of this PR that fixes the `make check` command.

This PR include

- adding the missing propriety in the listconfig.schema.json (https://github.com/ElementsProject/lightning/issues/4773)
- adding in the `make doc-all` the possibility to format the JSON schema, this change causes a lot of hash changes in the `.md` files sorry :(
- Fixed the make check-mainpages to skip the undocumented propriety. However, not all the experimental propriety are undocumented (e.g: offers) so I added a JSON array in `doc/undoc-flags.json` to track all the flags that don't need docs for now.
- The last point required to modify a little bit the `tools/check.mainpage.sh`

Hoping that this PR can help.

jsarenik commented:
ACK f3780bc9b

jsarenik commented:
@vincenzopalazzo Yes, using [`git rebase -i`](https://git-scm.com/docs/git-rebase#Documentation/git-rebase.txt--i) you can squash, reorder or remove, have a look at the [doc](https://git-scm.com/docs/git-rebase#Documentation/git-rebase.txt--i).

jsarenik commented:
@vincenzopalazzo The only missing checks are now the ACKbot and changelog.

For changelog (if there is none) add following to the git commit message (by running `git commit --amend`)

```

Changelog-None
```

I will ACK this as soon as `make check` works. Currently it end for me with this error:

```
$ make check
tor-service-password documented but not in schema!
make: *** [doc/Makefile:160: check-config-docs] Error 1
```

jsarenik commented:
Ah, actually [one CI check](https://github.com/ElementsProject/lightning/runs/4089915834?check_suite_focus=true) is rightfully failing.

```
... (line 124 follows)
tor-service*** but not in schema!
make: *** [doc/Makefile:160: check-config-docs] Error 1
Error: Process completed with exit code 2.
```

vincenzopalazzo commented:
Rebased on top of master and cleanup of the git history. This should contain also the fixes for the newest propriety added in version 0.10.2.

jsarenik commented:
ACK d95599187

jsarenik commented:
ACK d9559918725c309d61ce395d62f3f85229859985

whitslack commented on doc/lightningd-config.5.md:
```diff
@@ -212,6 +212,9 @@ If there is no `hsm_secret` yet, `lightningd` will create a new encrypted secret
 If you have an unencrypted `hsm_secret` you want to encrypt on-disk, or vice versa,
 see lightning-hsmtool(8).
 
+ **max-dust-htlc-exposure-msat**=*MILLISATOSHI*
+Option which limits the total amount of sats to be allowed as dust on a channel.
+
```
Doesn't this really belong in the **Lightning channel and HTLC options** section, perhaps adjacent to **`max-concurrent-htlcs`**?

vincenzopalazzo replied:
Mh, I agree with that thanks

jsarenik commented:
ACK 4af5942ac24ebc8ef472c06832b6138ea6bb6c66

jsarenik commented:
[The failed test](https://github.com/ElementsProject/lightning/runs/4109465137?check_suite_focus=true) looks like a false-positive

jsarenik reviewed (approved):
ACK 4af5942ac

andrewtoth commented:
> maybe we need to put jq in the install section of the doc?

I think this should be done here, yes. I had to install it when testing this PR (somehow I didn't have it installed already?).

andrewtoth commented on doc/schemas/WRITING_SCHEMAS.md:
```diff
@@ -39,6 +39,9 @@ You should always list all fields which are *always* present in
 We extend the basic types; see
 [contrib/pyln-testing/pyln/testing/fixtures.py](fixtures.py).
 
+In addition, before to commit a new schema or a new version of it, make sure that it
```
grammar nit:

In addition, before *committing* a new schema or a new version of it, make sure that it

vincenzopalazzo replied:
Thanks 😄

vincenzopalazzo commented:
>I think this should be done here, yes. I had to install it when testing this PR (somehow I didn't have it installed already?).

IMO Better had this dependence in the doc.

rustyrussell reviewed (approved):
Ack f89de630283edf6ef7969e797ccdf8758723250b

andrewtoth reviewed (approved):
ACK f89de630283edf6ef7969e797ccdf8758723250b

rustyrussell commented:
Trivial rebase on master (which had broken CI!).

Ack https://github.com/ElementsProject/lightning/commit/7fbc93b3612fee0466bb9733793de63439c629ba

vincenzopalazzo commented:
trivial rebase on actual master https://github.com/ElementsProject/lightning/commit/c1f534a159ad97eaaea1cd61bcc54b1e9bcd620c

rustyrussell commented:
Ack https://github.com/ElementsProject/lightning/commit/556369cc0d8ab2221fa18a157357ea4f6429307c
